File: dataStructure/BinarySearchTree.java
/**
 * BinarySearchTree
 * 
 * search, delete, insert => O(logN)
 */

package dataStructure;

public class BinarySearchTree {
    Node root;

    // Node Class
    class Node {
        int data;
        Node left;
        Node right;

        Node(int item) {
            this.data = item;
        }

        public String toString() {
            return "data: " + this.data;
        }
    }

    /**
     * Insert
     */
    public boolean insert(int item) {
        if (search(item) != null) {
            System.out.println("중복 데이터 " + item + " (이)가 있으므로 insert하지 않음.");
            return false;
        }

        if (root == null) {
            root = new Node(item);
            return true;
        } else {
            return insertNode(root, item);
        }
    }

    private boolean insertNode(Node root, int item) {
        if (root.data > item) {
            if (root.left != null)
                return insertNode(root.left, item);
            else
                root.left = new Node(item);
        } else {
            if (root.right != null)
                return insertNode(root.right, item);
            else
                root.right = new Node(item);
        }
        return true;
    }

    /**
     * Search
     */
    public Node search(int item) {
        if (root == null)
            return null;
        else 
            return searchNode(root, item);
    }

    private Node searchNode(Node root, int item) {
        if (root.data == item) {
            return root;
        } else if (root.data > item) {
            if (root.left != null)
                return searchNode(root.left, item);
        } else {
            if (root.right != null)
                return searchNode(root.right, item);
        }
        return null;
    }

    /**
     * Delete
     */
    public boolean delete(int item) {
        if (root == null) {
            System.out.println("트리가 존재하지 않습니다.");
            return false;
        }
        if (root.data == item) {  // 삭제하려는 데이터가 root 노드
            switch (childNum(root)) {
                case 0:
                    root = null;
                    break;
                case 1:
                    if (root.left == null)   // 오른쪽 자식만 있는 경우
                        root = root.right;
                    else                    // 왼쪽 자식만 있는 경우
                        root = root.left;
                    break;
                case 2:
                    Node[] successorInfo = getSuccessor(root);
                    Node parentOfSuccessor = successorInfo[0];
                    Node successor = successorInfo[1];

                    root.data = successor.data;             // successor의 '데이터'를 옮기고
                    if (parentOfSuccessor != root) {
                        parentOfSuccessor.left = successor.right; 
                    } else {
                        root.right = successor.right;
                    }
                    break;
            }
            return true;
        } else {
            if (root.data > item)
                return deleteNode(root, root.left, item);
            else
                return deleteNode(root, root.right, item);
        }
    }

    private boolean deleteNode(Node parent, Node curr, int item) {
        if (curr == null) {
            System.out.println("데이터가 존재하지 않습니다.");
            return false;
        }

        if (curr.data == item) {
            switch (childNum(curr)) {
                case 0:
                    if (parent.left == curr)
                        parent.left = null;
                    else
                        parent.right = null;
                    break;
                case 1:
                    if (parent.left == curr)
                        if (curr.left != null)
                            parent.left = curr.left;
                        else
                            parent.left = curr.right;
                    else
                        if (curr.left != null)
                            parent.right = curr.left;
                        else
                            parent.right = curr.right;
                    break;
                case 2:
                    // 자식이 둘이면 그 노드의 석세서와 석세서의 부모노드를 구하고 작업
                    Node[] successorInfo = getSuccessor(curr);
                    Node parentOfSuccessor = successorInfo[0];
                    Node successor = successorInfo[1];

                    curr.data = successor.data;             // successor의 '데이터'를 옮기고
                    if (parentOfSuccessor != curr) {
                        parentOfSuccessor.left = successor.right; 
                    } else {
                        curr.right = successor.right;
                    }
                    break;
            }
            return true;
        } else {
            if (curr.data > item)
                return deleteNode(curr, curr.left, item);
            else
                return deleteNode(curr, curr.right, item);
        }
    }

    /**
     * 파라미터 노드의 child node 수
     */
    private int childNum(Node root) {
        if (root.left == null && root.right == null)
            return 0;
        else if (root.left != null && root.right != null)
            return 2;
        else
            return 1;
    }

    /**
     * 파라미터 노드의 successor & successor의 부모노드를 Node[] 타입으로 return
     */
    private Node[] getSuccessor(Node root) {
        Node parent = root;
        root = root.right;

        while (root.left != null) {
            parent = root;
            root = root.left;
        }

        Node[] result = {parent, root};
        return result;
    }

    public static void main(String[] args) {
        /**
         * bst is ....
         *   10                           60
         *    \                           /
         *     50                       25
         *     /\                       /\
         *   20  60        ===>       17  30      
         *   /\                           
         * 17  30                       
         *     /
         *   25  
         */
        BinarySearchTree bst = new BinarySearchTree();
        bst.insert(10);
        bst.insert(10);  // prints "중복 데이터 10 (이)가 있으므로 insert하지 않음."
        bst.insert(50);
        bst.insert(20);
        bst.insert(30);
        bst.insert(25);
        bst.insert(17);
        bst.insert(60);

        System.out.println(bst.root.right.right);       // prints "data: 60"
        bst.delete(60);
        System.out.println(bst.root.right.right);       // prints "null"
        bst.insert(60);
        System.out.println(bst.root.right.right);       // prints "data: 60"

        System.out.println(bst.root.right.left);        // prints "data: 20"
        bst.delete(20);
        System.out.println(bst.root.right.left);        // prints "data: 25"
        System.out.println(bst.root.right.left.right);  // prints "data: 30"

        bst.delete(10);
        System.out.println(bst.root);                   // prints "data: 50"
        System.out.println(bst.search(50).left);        // prints "data: 25"

        System.out.println("==================================");
        bst.delete(50);
        System.out.println(bst.root);                   // prints "data: 60"
        System.out.println(bst.root.left);              // prints "data: 25"
        System.out.println(bst.root.left.left);         // prints "data: 17"
        System.out.println(bst.root.left.right);        // prints "data: 30"
        System.out.println(bst.root.right);             // prints "data: null"

        System.out.println("==================================");
        bst.delete(25);
        System.out.println(bst.root);                   // prints "data: 60"
        System.out.println(bst.root.left);              // prints "data: 30"
        System.out.println(bst.root.left.left);         // prints "data: 17"
        System.out.println(bst.root.left.right);        // prints "null"
        System.out.println(bst.root.right);             // prints "null"
    }
}

File: python/data_structure/binary_search_tree.py
import unittest
from typing import Optional
from abc import ABC, abstractmethod


class Node:
    def __init__(self, item):
        self.data = item
        self.left = None
        self.right = None

    def __str__(self):
        return 'data: ' + str(self.data)


class InnerBinarySearchTree(ABC):
    @abstractmethod
    def insert(self, item: int) -> bool: ...

    @abstractmethod
    def delete(self, item: int) -> bool: ...

    @abstractmethod
    def search(self, item: int) -> Node: ...

    @abstractmethod
    def preorder(self) -> list: ...

    @abstractmethod
    def inorder(self) -> list: ...

    @abstractmethod
    def postorder(self) -> list: ...


class BinarySearchTree(InnerBinarySearchTree):
    def __init__(self):
        self._root = None
        self.preorder_lst = []
        self.inorder_lst = []
        self.postorder_lst = []

    # class Node:
    #     def __init__(self, item):
    #         self.data = item
    #         self.left = None
    #         self.right = None

    #     def __str__(self):
    #         return 'data: ' + str(self.data)

    def insert(self, item: int) -> Optional[bool]:
        """Insert item in BST

        Keyword arguments:
            item: int
        Return:
            bool
        """
        if self.search(item) is not None:
            print('중복 데이터 ' + str(item) + '(이)가 있으므로 insert하지 않음.')
            return False

        if self._root is None:
            self._root = Node(item)
            return True
        else:
            self._insert_node(self._root, item)

    def _insert_node(self, root, item):
        if root.data > item:
            if root.left is not None:
                return self._insert_node(root.left, item)
            else:
                root.left = Node(item)
        else:
            if root.right is not None:
                return self._insert_node(root.right, item)
            else:
                root.right = Node(item)

        return True

    def delete(self, item: int) -> bool:
        """Delete item in BST

        Keyword arguments:
            item: int
        Return:
            bool
        """
        if self._root is None:
            print('트리가 존재하지 않습니다.')
            return False

        if self._root.data == item:
            child_num = self._child_num(self._root)
            if child_num == 0:
                self._root = None
            elif child_num == 1:
                if self._root.left is None:
                    self._root = self._root.right
                else:
                    self._root = self._root.left
            else:
                tmp = self.successor(self._root)
                parent_of_successor = tmp[0]
                successor = tmp[1]

                if parent_of_successor != self._root:
                    self._root.data = successor.data
                    parent_of_successor.left = successor.right
                else:
                    self._root.data = successor.data
                    self._root.right = successor.right
            return True
        else:
            if self._root.data > item:
                return self._delete_node(self._root, self._root.left, item)
            else:
                return self._delete_node(self._root, self._root.right, item)

    def _delete_node(self, parent, curr, item):
        if curr is None:
            print('데이터가 존재하지 않습니다.')
            return False

        if curr.data == item:
            child_num = self._child_num(curr)
            if child_num == 0:
                if parent.left == curr:
                    parent.left = None
                else:
                    parent.right = None
            elif child_num == 1:
                if parent.left == curr:
                    if curr.left is None:
                        parent.left = curr.right
                    else:
                        parent.left = curr.left
                else:
                    if curr.left is None:
                        parent.right = curr.right
                    else:
                        parent.right = curr.left
            else:
                tmp = self.successor(curr)
                parent_of_successor = tmp[0]
                successor = tmp[1]

                curr.data = successor.data
                if parent_of_successor != curr:
                    parent_of_successor.left = successor.right
                else:
                    curr.right = successor.right
            return True
        else:
            if curr.data > item:
                return self._delete_node(curr, curr.left, item)
            else:
                return self._delete_node(curr, curr.right, item)

    def _child_num(self, node):
        """Returns child_node num of node"""
        if node.left is None and node.right is None:
            return 0
        elif node.left is not None and node.right is not None:
            return 2
        else:
            return 1

    def successor(self, node):
        """Returns (parent_of_successor, successor)"""
        parent = node
        node = node.right

        while node.left is not None:
            parent = node
            node = node.left

        return parent, node

    def search(self, item: int) -> Optional[Node]:
        """Search item in BST

        Keyword arguments:
            item: intk
            Node
        """
        if self._root is None:
            return None
        else:
            return self._search_node(self._root, item)

    def _search_node(self, curr, item):
        if curr.data == item:
            return curr

        if curr.data > item:
            if curr.left is not None:
                return self._search_node(curr.left, item)
        else:
            if curr.right is not None:
                return self._search_node(curr.right, item)
        return None

    def preorder(self):
        if self._root is not None:
            self._preorder_traversal(self._root)

    def _preorder_traversal(self, curr):
        self.preorder_lst.append(curr.data)

        if curr.left is not None:
            self._preorder_traversal(curr.left)
        if curr.right is not None:
            self._preorder_traversal(curr.right)

    def inorder(self):
        if self._root is not None:
            self._inorder_traversal(self._root)

    def _inorder_traversal(self, curr):
        if curr.left is not None:
            self._inorder_traversal(curr.left)

        self.inorder_lst.append(curr.data)

        if curr.right is not None:
            self._inorder_traversal(curr.right)

    def postorder(self):
        if self._root is not None:
            self._postorder_traversal(self._root)

    def _postorder_traversal(self, curr):
        if curr.left is not None:
            self._postorder_traversal(curr.left)
        if curr.right is not None:
            self._postorder_traversal(curr.right)

        self.postorder_lst.append(curr.data)


class BinarySearchTreeTest(unittest.TestCase):
    def setUp(self):
        #   10
        #    \
        #     50
        #     /\
        #   20  60
        #   /\
        # 17  30
        #     /
        #   25
        self.bst = BinarySearchTree()
        self.bst.insert(10)
        self.bst.insert(10)  # prints "중복 데이터 10 (이)가 있으므로 insert하지 않음."
        self.bst.insert(50)
        self.bst.insert(20)
        self.bst.insert(30)
        self.bst.insert(25)
        self.bst.insert(17)
        self.bst.insert(60)

    def test_all(self):
        # bst is ...
        #   10                           60
        #    \                           /
        #     50                       25
        #     /\                       /\
        #   20  60        ===>       17  30
        #   /\
        # 17  30
        #     /
        #   25
        self.assertEqual(self.bst._root.right.right.data, 60)
        self.bst.delete(60)
        self.assertEqual(self.bst._root.right.right, None)
        self.bst.insert(60)
        self.assertEqual(self.bst._root.right.right.data, 60)

        self.assertEqual(self.bst._root.right.left.data, 20)
        self.bst.delete(20)
        self.assertEqual(self.bst._root.right.left.data, 25)
        self.assertEqual(
            self.bst._root.right.left.right.data, 30
        )

        self.bst.delete(10)
        self.assertEqual(self.bst._root.data, 50)
        self.assertEqual(self.bst.search(50).left.data, 25)

        self.bst.delete(50)
        self.assertEqual(self.bst._root.data, 60)
        self.assertEqual(self.bst._root.left.data, 25)
        self.assertEqual(self.bst._root.left.left.data, 17)
        self.assertEqual(self.bst._root.left.right.data, 30)
        self.assertIsNone(self.bst._root.right)

        self.bst.delete(25)
        self.assertEqual(self.bst._root.data, 60)
        self.assertEqual(self.bst._root.left.data, 30)
        self.assertEqual(self.bst._root.left.left.data, 17)
        self.assertIsNone(self.bst._root.left.right)
        self.assertIsNone(self.bst._root.right)

    def test_traversal(self):
        self.bst.preorder()
        self.assertEqual(self.bst.preorder_lst, [10, 50, 20, 17, 30, 25, 60])

        self.bst.inorder()
        self.assertEqual(self.bst.inorder_lst, [10, 17, 20, 25, 30, 50, 60])

        self.bst.postorder()
        self.assertEqual(self.bst.postorder_lst, [17, 25, 30, 20, 60, 50, 10])
